Mathlib/CategoryTheory/Limits/Preserves/Presheaf.lean

```diff
@@ -0,0 +1,179 @@
+
+This is one of the keys steps of establishing the equivalence `Ind C ≌ (Cᵒᵖ ⥤ₗ Type u)` for a
+*small* finitely cocomplete category `C`.
+
+## Implementation notes
+
+The theorem is also true for large categories and the proof given here generalizes to this case on
+paper. However, our infrastructure for commuting finite limits of shape `J` and filtered colimits
+of shape `K` is fundamentally not equipped to deal with the case where not all limits of shape `K`
+exist. In fact, not even the definition `colimitLimitToLimitColimit` can be written down if not
+all limits of shape `K` exist. Refactoring this to the level of generality we need would be a major
+undertaking. Since the application to the category of Ind-objects only require the case where `C`
+is small, we leave this as a TODO.
+
+## References
+* [M. Kashiwara, P. Schapira, *Categories and Sheaves*][Kashiwara2006], Proposition 3.3.13
+* [F. Borceux, *Handbook of Categorical Algebra 1*][borceux-vol1], Proposition 6.1.2
+-/
+
+open CategoryTheory Limits
+
+universe v u
+
+section LargeCategory
+
+variable {C : Type u} [Category.{v} C] [HasFiniteColimits C] (A : Cᵒᵖ ⥤ Type v)
+
+/-- If `C` is a finitely cocomplete category and `A : Cᵒᵖ ⥤ Type u` is a presheaf that preserves
+finite limites, then `CostructuredArrow yoneda A` is filtered.
+
+One direction of Proposition 3.3.13 of [Kashiwara2006].
+-/
+theorem isFiltered_costructuredArrow_yoneda_of_preservesFiniteLimits
+    [PreservesFiniteLimits A] : IsFiltered (CostructuredArrow yoneda A) := by
+  suffices IsFiltered A.Elementsᵒᵖ from
+    IsFiltered.of_equivalence (CategoryOfElements.costructuredArrowYonedaEquivalence _)
+  suffices IsCofiltered A.Elements from isFiltered_op_of_isCofiltered A.Elements
+  suffices HasFiniteLimits A.Elements from IsCofiltered.of_hasFiniteLimits A.Elements
+  exact ⟨fun J _ _ => inferInstance⟩
+
+end LargeCategory
+
+variable {C : Type u} [SmallCategory C] [HasFiniteColimits C]
+
+namespace CategoryTheory.Limits
+
+variable (A : Cᵒᵖ ⥤ Type u)
+
+namespace PreservesFiniteLimitsOfIsFilteredCostructuredArrowYonedaAux
+
+variable {J : Type} [SmallCategory J] [FinCategory J] (K : J ⥤ Cᵒᵖ)
+
+/-- (Implementation) This is the bifunctor we will apply "filtered colimits commute with finite
+limits" to. -/
+def functorToInterchange : J ⥤ CostructuredArrow yoneda A ⥤ Type u :=
+  K ⋙ coyoneda ⋙ (whiskeringLeft _ _ _).obj (CostructuredArrow.proj _ _)
+
+/-- (Implementation) The definition of `functorToInterchange`, up to functor associativity. -/
+@[simps!]
+def functorToInterchangeIso : functorToInterchange A K ≅
+    K ⋙ (coyoneda ⋙ (whiskeringLeft _ _ _).obj (CostructuredArrow.proj _ _)) :=
+  NatIso.ofComponents (fun _ => Iso.refl _)
+
+/-- (Implementation) One way to express the flipped version of our functor. -/
+@[simps!]
+def flipFunctorToInterchange : (functorToInterchange A K).flip ≅
+    ((CostructuredArrow.proj yoneda A ⋙ yoneda) ⋙ (whiskeringLeft J Cᵒᵖ (Type u)).obj K) :=
```

Comment:
Is there a particular reason you prefer to associate the composition on the left?

Reply:
This is because the type of `Presheaf.tautologicalCocone` is `Cocone (CostructuredArrow.proj yoneda P ⋙ yoneda)`.